src/lib/asl.c

```diff
@@ -0,0 +1,44 @@
+#include <asl.h>
+#include <pwd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <time.h>
+
+#include <SystemConfiguration/SystemConfiguration.h>
+
+#include <xhyve/asl.h>
+
+static aslclient asl = NULL;
+static aslmsg log_msg = NULL;
+
+static unsigned char buf[4096];
+/* Index of the _next_ character to insert in the buffer. */
+static size_t buf_idx = 0;
+
+/* asl is deprecated in favor of os_log starting with macOS 10.12.  */
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+/* Initialize ASL logger and local buffer. */
+void asl_init(void)
+{
+	asl = asl_open(NULL, NULL, 0);
+	log_msg = asl_new(ASL_TYPE_MSG);
+}
+
+
+/* Send one character to the logger: wait for full lines before actually sending. */
+void asl_put(uint8_t c)
+{
+	if ((c == '\n') || (c == 0)) {
+		buf[buf_idx] = 0;
+		asl_log(asl, log_msg, ASL_LEVEL_NOTICE, "%s", buf);
+		buf_idx = 0;
+	} else {
+		if (buf_idx + 2 >= sizeof(buf)) {
+			/* Running out of space, flush.  */
```

Comment:
Well, `asl_put('\n')` flushes.  When should I start to help us see it?

Reply:
but then you start writing over the end of the buffer...

Reply:
It took me a moment to grok that too. How about adding `asl_flush()` with the contents of the `\n` case and calling that from both?

Reply:
@ijc Ok, good idea.  @rn sorry, I don't understand what you mean.

Reply:
@rn I think `buf_idx + 2 >= sizeof(buf)` ensures there is enough space for the `\0` which would be added for flushing. I'm not quite sure why it is +2 rather than +1 though, with the use of `>=` it seems like it would flush a byte too soon (but, meh).

Reply:
@ijc just pointed out the "subtle" recursive call to `asl_put()`. This is just plain crazy! don't do that.

Reply:
I suppose intruding on the `asl_*` namespace also used by the OS provided functions is moderately confusing. Perhaps our wrappers should all be `xhyve_asl_*` to avoid that?